packages/hardhat-zksync-verify/src/solc/bytecode.ts

```diff
@@ -0,0 +1,137 @@
+        };
+    }
+
+    public getInferredSolcVersion(): string {
+        return this._version;
+    }
+
+    public getExecutableSection(): string {
+        const { start, length } = this._executableSection;
+        return this._bytecode.slice(start, length);
+    }
+
+    public hasMetadata(): boolean {
+        return this._metadataSection.length > 0;
+    }
+}
+
+export async function extractMatchingContractInformation(
+    sourceName: SourceName,
+    contractName: ContractName,
+    buildInfo: BuildInfo,
+    deployedBytecode: Bytecode
+): Promise<ContractInformation | null> {
+    const contract = buildInfo.output.contracts[sourceName][contractName];
+
+    if (contract.hasOwnProperty('evm')) {
+        const { bytecode: runtimeBytecodeSymbols } = contract.evm;
+
+        const analyzedBytecode = runtimeBytecodeSymbols
+            ? await compareBytecode(deployedBytecode, runtimeBytecodeSymbols)
+            : null;
+
+        if (analyzedBytecode !== null) {
+            return {
+                ...analyzedBytecode,
+                compilerInput: buildInfo.input,
+                compilerOutput: buildInfo.output,
+                solcVersion: buildInfo.solcVersion,
+                sourceName,
+                contractName,
+                contract,
+            };
+        }
+
+        return null;
+    }
+
+    return null;
+}
+
+export async function compareBytecode(
+    deployedBytecode: Bytecode,
+    runtimeBytecodeSymbols: CompilerOutputBytecode
+): Promise<BytecodeExtractedData | null> {
+    // We will ignore metadata information when comparing. Etherscan seems to do the same.
+    const deployedExecutableSection = deployedBytecode.getExecutableSection();
+    const runtimeBytecodeExecutableSectionLength = runtimeBytecodeSymbols.object.length;
+
+    if (deployedExecutableSection.length !== runtimeBytecodeExecutableSectionLength) {
+        return null;
+    }
+
+    // Normalize deployed bytecode according to this contract.
+    const { normalizedBytecode } = await normalizeBytecode(deployedExecutableSection);
+
+    const { normalizedBytecode: referenceBytecode } = await normalizeBytecode(runtimeBytecodeSymbols.object);
+
+    if (
+        normalizedBytecode.slice(0, deployedExecutableSection.length) ===
+        referenceBytecode.slice(0, deployedExecutableSection.length)
+    ) {
+        // The bytecode matches
+        return {
+            normalizedBytecode,
+        };
+    }
+
+    return null;
+}
+
+export async function normalizeBytecode(bytecode: string): Promise<BytecodeExtractedData> {
+    const nestedSliceReferences: NestedSliceReferences = [];
+
+    const addressSize = 20;
+    const push20OpcodeHex = '73';
+    const pushPlaceholder = push20OpcodeHex + '0'.repeat(addressSize * 2);
+    if (bytecode.startsWith(pushPlaceholder) && bytecode.startsWith(push20OpcodeHex)) {
+        nestedSliceReferences.push([{ start: 1, length: addressSize }]);
+    }
```

Comment:
If necessary we are normalizing a library object by replacing a PUSH20 instruction (opcode) with the current address so it can be directly compared with the deployed code.
I will add a comment to explain it further.

Reply:
Ok, but why is there 2 `bytecode.startsWith` calls in one `if`? Isn't the second one redundant, since `pushPlaceholder` starts with `push20OpcodeHex`?

Reply:
This was changed in the latest commit.